exchanges/poloniex/poloniex_websocket.go

```diff
@@ -28,1128 +31,497 @@ import (
-	if !ok {
-		return fmt.Errorf("%w original amount not string", errTypeAssertionFailure)
-	}
-	origOrderAmount, err := strconv.ParseFloat(origAmount, 64)
-	if err != nil {
+func (e *Exchange) processResponse(result *SubscriptionResponse, instance any) error {
+	if err := json.Unmarshal(result.Data, instance); err != nil {
 		return err
 	}
-
-	// null returned so ok check is not needed
-	clientOrderID, _ := notification[8].(string)
-	e.Websocket.DataHandler <- &order.Detail{
-		Exchange:        e.Name,
-		Price:           orderPrice,
-		RemainingAmount: orderAmount,
-		ExecutedAmount:  origOrderAmount - orderAmount,
-		Amount:          origOrderAmount,
-		OrderID:         strconv.FormatFloat(orderID, 'f', -1, 64),
-		Type:            order.Limit,
-		Side:            orderSide,
-		Status:          order.New,
-		AssetType:       asset.Spot,
-		Date:            timeParse,
-		Pair:            pair,
-		ClientOrderID:   clientOrderID,
-	}
+	e.Websocket.DataHandler <- instance
 	return nil
 }
 
-func (e *Exchange) processAccountBalanceUpdate(ctx context.Context, notification []any) error {
-	if len(notification) < 4 {
-		return errNotEnoughData
-	}
-
-	currencyID, ok := notification[1].(float64)
-	if !ok {
-		return fmt.Errorf("%w currency ID not float64", errTypeAssertionFailure)
-	}
-	curr, err := e.details.GetCode(currencyID)
+func (e *Exchange) handleSubscription(operation string, s *subscription.Subscription) (*SubscriptionPayload, error) {
+	pairFormat, err := e.GetPairFormat(asset.Spot, true)
 	if err != nil {
-		return err
-	}
-
-	walletType, ok := notification[2].(string)
-	if !ok {
-		return fmt.Errorf("%w wallet addr not string", errTypeAssertionFailure)
-	}
-
-	a, ok := notification[3].(string)
-	if !ok {
-		return fmt.Errorf("%w amount not string", errTypeAssertionFailure)
+		return nil, err
 	}
-	amount, err := strconv.ParseFloat(a, 64)
-	if err != nil {
-		return err
+	input := &SubscriptionPayload{
+		Event:   operation,
+		Channel: []string{strings.ToLower(s.QualifiedChannel)},
 	}
 
-	bal := accounts.Balance{
-		Currency: curr,
-		Total:    amount,
-		Free:     amount,
+	if len(s.Pairs) != 0 && s.QualifiedChannel != channelFuturesAccount {
+		input.Symbols = s.Pairs.Format(pairFormat).Strings()
 	}
 
-	id := deriveWalletType(walletType)
-	subAccts := accounts.SubAccounts{accounts.NewSubAccount(asset.Spot, id)}
-	subAccts[0].Balances.Set(curr, bal)
-
-	if err := e.Accounts.Save(ctx, subAccts, true); err != nil {
-		return err
+	switch s.Channel {
+	case subscription.OrderbookChannel, channelBooks:
+		input.Depth = int64(s.Levels) // supported orderbook levels are 5, 10 and 20
+	case channelCurrencies:
+		for _, p := range s.Pairs {
+			if !slices.Contains(input.Currencies, p.Base.String()) {
+				input.Currencies = append(input.Currencies, p.Base.String())
+			}
+		}
+	case subscription.MyOrdersChannel:
+		if s.Asset == asset.Spot && len(input.Symbols) == 0 {
+			input.Symbols = []string{"all"}
+		}
```

Comment:
This is all fine. Good work.

Just to note that the intention of templating is to *not* do this here.

In this case, I'd have made QualifiedChannel return the full Json, and then just set the Event on that. A few other exchanges do that.

But what you've done works, and I don't think we should change it just for consistency.
This doesn't *need* the "fan out" pattern that drove the templating to behave that way.